include/drivers/kscan.h

```diff
@@ -44,6 +50,31 @@ typedef void (*kscan_callback_t)(const struct device *dev, uint32_t row,
 				 uint32_t column,
 				 bool pressed);
 
+/**
+ * @brief Keyboard scan ISR callback called when user presses to trigger the
+ * keybaord scan interrupt.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ */
+typedef void (*kscan_isr_callback_t)(const struct device *dev);
+
+/**
+ * @brief Keyboard matrix configuration.
+ *
+ * This defines keyboard matrix configuration used by keyboard scan task and driver.
+ */
+struct kscan_matrix_config {
+	/*
+	 * The callback is called when the keyboard matrix poll task detects a col and row
+	 * release/press event. It is registered when kscan_config_matrix API is called.
+	 * It is the link to the user/application callback which is registered when the
+	 * kscan_config API is called by the application.
+	 */
+	kscan_callback_t callback;
+	kscan_isr_callback_t isr_callback;
+	atomic_t *enable_scan;
```

Comment:
so why do we need to types of callbacks? Where is this documented?

Reply:
I don't see a need for the `isr_callback`.  The only user is `kscan_matrix_poll.c`.  You could have the drivers  directly notify the matrix task.

```
static void kscan_npcx_ksi_isr(const struct device *dev, struct npcx_wui *wui)
{
	ARG_UNUSED(wui);

	if (IS_ENABLED(CONFIG_KCAN_MATRIX_TASK)) {
		kscan_matrix_scan_complete(dev);
	}
}
```

Alternately, you could pass a reference to the semaphore to the driver.

```
	if (data->config->matrix_task_sem) {
		k_sem_give(data->config->matrix_task_sem);
	}
```